### PR TITLE
JSDK-21: Custom File Helper for Mapping from a Local Path to a DS3 Path

### PR DESCRIPTION
**Pull request by DenverM80**
Create helper functions for differing local and remote paths.  Add sample file and unit tests.

### CODE REVIEW

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers.java [hunk 1]
```diff
@@ -203,5 +205,34 @@ public static Ds3ClientHelpers wrap(final Ds3Client client) {
...
+     */
+    public abstract Iterable<Ds3Object> removePrefixFromDs3ObjectsList(final Iterable<Ds3Object> objectsList, final String prefix);
+
+    /**
+     * Strip prefix from the beginning of objectName.  If objectName does not start with prefix, return objectName unmodified.
+     * @param objectName
+     * @param prefix
+     * @return
+     */
+    public static String stripLeadingPath( final String objectName, final String prefix) {
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers.java]: You have an extra space behind the '('

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers.java [hunk 2]
```diff
@@ -203,5 +205,34 @@ public static Ds3ClientHelpers wrap(final Ds3Client client) {
...
+        String returnString = objectName;
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers.java]: I might change this to be `if (objectName.startsWith(prefix)) { return objectName.substring(prefix.length()); } return objectName`.  It doesn't require allocating a new variable.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/LoggingFileObjectPutter.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+
+public class LoggingFileObjectPutter implements Ds3ClientHelpers.ObjectChannelBuilder {
+    final private FileObjectPutter objectPutter;
+    final private static Logger LOG = LoggerFactory.getLogger(LoggingFileObjectPutter.class);
+
+    public LoggingFileObjectPutter(final Path inputDirectory) {
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/LoggingFileObjectPutter.java]: If we really want this to be extensible we should change it so that the other object putter is passed in, instead of it being created inside of this class.  That will allow it be to used with other `FileObjectPutter` classes.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/LoggingFileObjectPutter.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+            this.objectPutter = new FileObjectPutter(inputDirectory);
+    }
+
+    @Override
+    public SeekableByteChannel buildChannel(final String s) throws IOException {
+        LOG.info("Putting " + s + " to ds3 endpoint");
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/LoggingFileObjectPutter.java]: If we make the change above, this message should be changed to say 'Opening' rather than 'Putting'

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectGetter.java [hunk 5]
```diff
@@ -38,7 +38,6 @@ public FileObjectGetter(final Path root) {
         this.root = root;
     }
 
-    @Override
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectGetter.java]: Why was this removed?

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectPutter.java [hunk 6]
```diff
@@ -37,7 +37,6 @@ public FileObjectPutter(final Path root) {
         this.root = root;
     }
 
-    @Override
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectPutter.java]: Why was this removed?

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java [hunk 7]
```diff
@@ -24,12 +24,12 @@
 import com.spectralogic.ds3client.models.ListBucketResult;
 import com.spectralogic.ds3client.models.bulk.*;
 import com.spectralogic.ds3client.serializer.XmlProcessingException;
+import org.slf4j.Logger;
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java]: Where is the logger being used?  I don't see any code additions below that use it.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    public PrefixRemoverObjectChannelBuilder(final Ds3ClientHelpers.ObjectChannelBuilder channelBuilder, final String prefix) {
+        this.channelBuilder = channelBuilder;
+        this.prefix = prefix;
+    }
+
+    @Override
+    public SeekableByteChannel buildChannel(final String key) throws IOException {
+        final String strippedKey = Ds3ClientHelpers.stripLeadingPath(key, prefix);
+
+        Files.createDirectories(Paths.get(strippedKey).getParent());
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java]: Why is this ObjectChannelBuilder creating the directory?  Shouldn't it be the wrapped ObjectChannelBuilder that's creating the directory?

File: ds3-sdk/src/test/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers_Test.java [hunk 9]
```diff
@@ -507,4 +507,53 @@ private static Objects chunk2(final boolean inCache) {
...
+    @Test
+    public void testAddPrefixToDs3ObjectsList() throws IOException, SignatureException {
+        final List<Ds3Object> ds3ObjectList = Lists.newArrayList(new Ds3Object("foo"), new Ds3Object("bar"));
+        final Ds3ClientHelpers helper = Ds3ClientHelpers.wrap(buildDs3ClientForBulk());
+        final Iterable<Ds3Object> modifiedDs3ObjectList = helper.addPrefixToDs3ObjectsList(ds3ObjectList, "baz/");
+
+        boolean foundBazFoo = false;
+        boolean foundBazBar = false;
+
+        for (Ds3Object obj : modifiedDs3ObjectList) {
```
Comment (rpmoore) [inline ds3-sdk/src/test/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers_Test.java]: Missing final

File: ds3-sdk/src/test/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers_Test.java [hunk 10]
```diff
@@ -507,4 +507,53 @@ private static Objects chunk2(final boolean inCache) {
...
+    @Test
+    public void testRemovePrefixToDs3ObjectsList() throws IOException, SignatureException {
+        final List<Ds3Object> ds3ObjectList = Lists.newArrayList(new Ds3Object("foo/bar"), new Ds3Object("foo/baz"));
+        final Ds3ClientHelpers helper = Ds3ClientHelpers.wrap(buildDs3ClientForBulk());
+        final Iterable<Ds3Object> modifiedDs3ObjectList = helper.removePrefixFromDs3ObjectsList(ds3ObjectList, "foo/");
+
+        boolean foundBar = false;
+        boolean foundBaz = false;
+
+        for (Ds3Object obj : modifiedDs3ObjectList) {
```
Comment (rpmoore) [inline ds3-sdk/src/test/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpers_Test.java]: missing final

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectPutter.java (hunk 6 above)
Comment (DenverM80) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/FileObjectPutter.java]: This function isn't actually overriding anything - its implementing the interface

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java (hunk 7 above)
Comment (DenverM80) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java]: hmm.  I must've had it at one point, and then pulled it out.  I'll update.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java (hunk 8 above)
Comment (DenverM80) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java]: Yes, this is now redundant.  Will update.

Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java]: It doesn't look like you removed this.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+
+/**
+ * Writes files to the local file system with a different path by first stripping off a remote prefix
+ * and then adding a local prefix.
+ */
+public class PrefixRemoverObjectChannelBuilder implements ObjectChannelBuilder {
+    final private Ds3ClientHelpers.ObjectChannelBuilder channelBuilder;
+    private String prefix;
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/PrefixRemoverObjectChannelBuilder.java]: Missing final?

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java [hunk 12]
```diff
@@ -206,4 +203,21 @@ public FileVisitResult visitFile(final Path file, final BasicFileAttributes attr
         });
         return objects;
     }
+
+    public Iterable<Ds3Object> addPrefixToDs3ObjectsList(final Iterable<Ds3Object> objectsList, final String prefix) {
+        final List<Ds3Object> newObjectsList = new ArrayList<>();
+        for ( final Ds3Object object: objectsList ) {
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java]: There is extra spaces on the inside of the parens.

File: ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java [hunk 13]
```diff
@@ -206,4 +203,21 @@ public FileVisitResult visitFile(final Path file, final BasicFileAttributes attr
...
+            final Ds3Object tmpObj = new Ds3Object( prefix + object.getName(), object.getSize());
+            newObjectsList.add(tmpObj);
+        }
+        return newObjectsList;
+    }
+
+    public Iterable<Ds3Object> removePrefixFromDs3ObjectsList(final Iterable<Ds3Object> objectsList, final String prefix) {
+        final List<Ds3Object> newObjectsList = new ArrayList<>();
+        for ( final Ds3Object object: objectsList ) {
```
Comment (rpmoore) [inline ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/Ds3ClientHelpersImpl.java]: Same thing here about the spaces.
